[MVC - Step 1] 션 미션 제출합니다.

os94 opened the pull request:
안녕하세요! 션 MVC 1단계 PR 제출드립니다!
홍시님께 피드백을 처음받아봐서 무척 설렙니다! 감사합니다 😃

os94 committed:
test: Reflection을 이용한 클래스 정보 출력

os94 committed:
test: Reflection을 이용한 test로 시작하는 메소드 실행

os94 committed:
test: Reflection을 이용한 @Test 애노테이션 메소드 실행

os94 committed:
test: Reflection을 이용한 private field 값 할당

os94 committed:
test: Reflection을 이용한 인자를 가진 생성자의 인스턴스 생성

os94 committed:
test: Reflection을 이용한 component scan

os94 committed:
feat: 애노테이션기반의 handler mapping 구현

os94 committed:
feat: 레거시 MVC와 애노테이션 기반 MVC 일부 통합

os94 committed:
chore: auto formatting

os94 committed:
refactor: cleanup code

os94 committed:
refactor: HandlerMapping 클래스들이 공통 인터페이스 구현하도록 통일

Laterality committed:
chore: update comment

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/DispatcherServlet.java:
```diff
@@ -11,39 +14,55 @@
+                mav.getView().render(mav.getModel(), req, resp);
+            } else if (handler instanceof Controller) {
+                String viewName = ((Controller) handler).execute(req, resp);
+                move(viewName, req, resp);
```
위 2개의 if문에서는 아래와 같은 공통된 흐름으로 로직이 작성돼 있습니다.

1. 조건 확인
2. 만족하는 타입의 메서드 실행

이를 기반으로 공통된 인터페이스를 만들어 분기문을 제거해보면 어떨까요?

os94 replied:
해당 로직을 인터페이스로 분리해, 구현체로 로직을 수행하고
service 메소드에선 일관된방식으로 처리할수있도록 수정할수있었습니다.

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/DispatcherServlet.java:
```diff
@@ -11,39 +14,55 @@
+    // return Controller or HandlerExecution
+    private Object selectHandler(HttpServletRequest req) {
+        return Optional.ofNullable(legacyHandlerMapping.getHandler(req))
+                .orElseGet(() -> annotationHandlerMapping.getHandler(req));
```
이 코드는 2가지 외에 새로운 `HandlerMapping` 이 추가되면 로직 변경이 필요합니다. 그리고 인스턴스 변수도 하나 더 추가해야 합니다.
현재 2가지 모두 같은 인터페이스를 구현하고 있습니다. 이를 활용해 새로운 타입이 추가되더라도 변경을 최소화할 수 있는 구조로 바꿔보면 어떨까요?

os94 replied:
`HandlerMapping`들을 리스트로 관리하도록 수정하였습니다 😃

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -15,11 +28,42 @@ public AnnotationHandlerMapping(Object... basePackage) {
+    private HandlerExecution invoke(Method method) {
+        return (request, response) -> {
+            try {
+                Object instance = method.getDeclaringClass().getConstructor().newInstance();
```
함수형 인터페이스가 실행될 때마다 매번 인스턴스를 생성합니다. 이러한 구조는 비효율적이지 않을까요?
인스턴스를 최초 한 번만 생성하도록 바꿔보면 어떨까요?

os94 replied:
함수형 인터페이스를 사용하며, `Object instance`가 매번 생성될건 미처 생각치못했던것같습니다!
인터페이스에 대한 구현체 클래스를 만들어, Map에 저장해 사용하도록 수정해보았습니다.

hongsii reviewed (changes requested):
안녕하세요 션!
요구사항 및 전반적인 객체 구현을 잘 하셨습니다.   💯 
몇 가지 피드백을 드렸으니, 확인하셔서 반영해보세요 :)

os94 committed:
refactor: HandlerExecutionImpl 클래스 추가

- AnnotationHandlerMapping.invoke()에서 매번 Object instance가 생성되는 구조를, 구현체 클래스를 추가하고 Map에 저장해서 사용하도록 개선

os94 committed:
refactor: DispatcherServlet의 HandlerMapping을 List로 관리하도록 수정

os94 committed:
refactor: HandlerAdapter 클래스를 추가해, DispatcherServlet.service() 개선

- HandlerAdapter 클래스를 추가해, DispatcherServlet.service()의 if문 분기를 없애고 List로 관리

os94 committed:
chore: exception 처리

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/DispatcherServlet.java:
```diff
@@ -1,57 +1,74 @@
 
-    private HandlerMapping rm;
+    private static final List<HandlerMapping> handlerMappings = new ArrayList<>();
+    private static final List<HandlerAdapter> handlerAdapters = new ArrayList<>();
```
`static` 으로 선언하신 이유가 있을까요?

os94 replied:
수정했습니다!^^;

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerAdapter.java:
```diff
@@ -0,0 +1,16 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class AnnotationHandlerAdapter implements HandlerAdapter {
```
추상화 👍

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/HandlerExecutionImpl.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.Map;
+import java.util.Objects;
+
+public class HandlerExecutionImpl implements HandlerExecution {
```
`Impl` 이란 이름은 좋은 네이밍이 아닙니다. 의미가 드러나는 명확한 네이밍으로 바꿔보면 어떨까요?

* [Naming the Java Implementation Classes](https://web.archive.org/web/20130331071928/http://isagoksu.com/2009/development/java/naming-the-java-implementation-classes)

os94 replied:
리팩토링 과정에서
기존의 `HandlerExecution` 함수형인터페이스를 제거하고
`HandlerExecutionImpl`의 네이밍을 `HandlerExecution`으로 수정하게 되었습니다 :)

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/HandlerExecutionImpl.java:
```diff
@@ -0,0 +1,79 @@
+
+    public Method getMethod() {
+        return method;
+    }
```
안쓰는 메서드는 삭제하는게 좋습니다!

os94 replied:
감사합니다 !

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/DispatcherServlet.java:
```diff
@@ -1,57 +1,74 @@
-        rm.initialize();
+    public void init() {
+        handlerMappings.stream()
+                .forEach(handlerMapping -> handlerMapping.initialize());
```
단순히 순회만 하는 코드라면 `Iterable.forEach(...)` 를 사용해도 좋을 듯 해요

* [What is difference between Collection.stream().forEach() and Collection.forEach()?](https://stackoverflow.com/questions/23218874/what-is-difference-between-collection-stream-foreach-and-collection-foreach)

os94 replied:
감사합니다! 알려주신 링크를 통해 두 방식의 차이점을 알게되었습니다!! 👍

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/DispatcherServlet.java:
```diff
@@ -1,57 +1,74 @@
+                logger.error(e.getMessage());
+                throw new RuntimeException(e);
+            }
+        });
```
람다 코드의 길이가 길어서 `service(...)` 메서드 로직을 한 눈에 파악하기 쉽지 않은 듯 해요. 리팩토링해보면 어떨까요?

os94 replied:
이어지는 피드백과 함께 수정했습니다.
코드가 훨씬 깔끔해진것 같습니다!! 😃

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/DispatcherServlet.java:
```diff
@@ -1,57 +1,74 @@
+        return handlerMappings.stream()
+                .map(handlerMapping -> handlerMapping.getHandler(req))
+                .filter(handler -> handler != null)
+                .findAny();
```
여기서 처리하지 않고 `Object` 가 아닌 `Optional<Object>` 로 반환하신 이유가 있을까요?

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -15,11 +27,34 @@ public AnnotationHandlerMapping(Object... basePackage) {
+    }
 
+    private HandlerExecution invoke(Method method) {
+        return (request, response) -> HandlerExecutionImpl.of(method).handle(request, response);
```
이미 인터페이스를 구현한 클래스인데, 이를 불필요하게 람다로 한 번 더 감싸주지 않아도 될 듯 합니다.

``` java
private HandlerExecution invoke(Method method) {
        return HandlerExecutionImpl.of(method);
}
```

현재는 람다로 감싸고 있기 떄문에 `HandlerExecutionImpl.of(...)` 메서드의 캐싱 로직이 유용했는데, 람다로 감싸지 않으면 중복된 데이터가 두 군데(`AnnotationHandlerMapping`, `HandlerExecutionImpl`)에 존재하기 때문에 `HandlerExecutionImpl.of(...)` 의 내부 캐싱 로직을 제거해도 될 것 같습니다.

os94 replied:
수정했습니다! 군데군데 놓친곳이 많았네요..ㅠ

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/HandlerExecutionImpl.java:
```diff
@@ -0,0 +1,79 @@
+
+    private HandlerExecutionImpl(Method method) {
+        try {
+            this.declaredObject = method.getDeclaringClass().getConstructor().newInstance();
```
하나의 클래스에 10개의 `@RequestMapping` 메서드가 있으면 10번의 인스턴스가 생성되지 않을까요?
외부에서 인스턴스를 미리 생성해두고 필요할 때 마다 주입해주면 불필요하게 여러 번 생성하지 않아도 될 것 같아요 :)

os94 replied:
`ControllerScanner`를 추가해보았습니다! :)

hongsii reviewed (changes requested):
피드백 반영 및 추상화를 잘 하셨네요! 💯 
추가로 몇 가지 개선점이 보여 피드백을 드렸으니, 확인하시고 반영해주세요 :)

os94 committed:
chore: package 정리

os94 committed:
refactor: Rename HandlerExecutionImpl -> HandlerMethod

os94 committed:
refactor: DispatcherServlet.service() 리팩토링

os94 committed:
refactor: 불필요한 익명함수 및 캐싱로직 제거 (HandlerMethod, AnnotationHandlerMapping)

os94 committed:
refactor: ControllerScanner 추가

- AnnotationHandlerMapping의 역할을 분리해 ControllerScanner 클래스 생성
- HandlerExecution을 함수형인터페이스에서 구현체 클래스로 대체

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -1,25 +1,41 @@
 
+    private void mapControllerMethod(Method method) {
+        RequestMapping requestMapping = method.getAnnotation(RequestMapping.class);
+        RequestMethod[] methods = requestMapping.method();
+        if (requestMapping.method().length == 0) {
+            methods = RequestMethod.values();
+        }
```
메서드로 추출하면 좀 더 로직의 가독성이 높아질 수 있을 듯 합니다 :)

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/ControllerScanner.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ControllerScanner {
```
클래스 추출 👍

hongsii commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/handleradapter/HandlerAdapter.java:
```diff
@@ -0,0 +1,12 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public interface HandlerAdapter {
```
추상화 💯

hongsii reviewed (approved):
피드백 반영을 잘해주셨습니다 💯 
간단한 피드백 몇 가지 드렸으니, 다음 단계에서 같이 반영해보세요 :)
